torch/fx/experimental/symbolic_shapes.py

```diff
@@ -269,20 +326,68 @@ def eval(cls, base, divisor):
     'sym_sqrt': lambda a: sympy.sqrt(a),
 }
 
+sizes_strides_methods = {
+    'is_non_overlapping_and_dense': lambda *args: IsNonOverlappingAndDenseIndicator(*args),
+}
+
+def eval_is_non_overlapping_and_dense(sizes, strides):
+    dim = len(sizes)
+
+    # Short-circuits for tensors of rank one, which are
+    # non-overlapping and "dense" if their stride is one
+    if dim == 1:
+        return strides[0] == 1
+
+    # Checks that there exists a permutation of the strides s.t. the tensor would be contiguous
+    # Sorts (length, stride) pairs by stride
+    lengths_and_strides = sorted(
+        tuple(zip(sizes, strides)), key=operator.itemgetter(1)
+    )
+
+    expected_stride = 1
+    for length, stride in lengths_and_strides:
+
+        if length == 1:
+            continue
+
+        if stride != expected_stride:
+            return False
+
+        expected_stride *= length
+
+    return True
+
+def is_non_overlapping_and_dense(sizes, strides):
+    base = None
+    for s in itertools.chain(sizes, strides):
+        if isinstance(s, SymInt):
+            base = s
+            break
+
+    assert base is not None
+    return wrap_node(base.node.is_non_overlapping_and_dense(
+        [to_node(base.node, s) for s in sizes],
+        [to_node(base.node, s) for s in strides],
+    ))
+
 unary_magic_methods = {
     'sym_float',
     'ceil',
     'floor',
     'neg',
     'sym_sqrt',
+    'sym_not',
 }
 
+bool_magic_methods = {"and", "or", "sym_not"}
+
 magic_methods_on_math = {"ceil", "floor"}
-magic_methods_on_submodule = {"sym_float", "sym_sqrt", "sym_min", "sym_max"}
+magic_methods_on_submodule = {"sym_float", "sym_sqrt", "sym_min", "sym_max", "sym_not"}
+magic_methods_on_operator_with_trailing_underscore = {"and", "or"}
```

Comment:
Why not call them and_ and or_ directly?
Also are these actually inplace?

Reply:
because when I double-underscore them they need to be `__and__`. So one way or another I need to either remove/add the underscore.